evm/src/main/java/org/hyperledger/besu/evm/operation/BlockHashOperation.java

```diff
@@ -29,45 +32,105 @@
 /** The Block hash operation. */
 public class BlockHashOperation extends AbstractFixedCostOperation {
 
-  private static final int MAX_RELATIVE_BLOCK = 255;
+  /** The HISTORY_STORAGE_ADDRESS */
```

Comment:
I really don't like this implementation, as it forces design decisions onto downstream users of the library.

This logic should live in a new BlockHashLookup function that is wired in in the mainnet protocol schedule, rather than re-writing the operation based on a flag.

Reply:
Here's a WIP of how I think it should work - https://github.com/hyperledger/besu/compare/main...shemnon:besu:eip-2935-danno

* Move all block hash lookup out of the operation, the operation only delegates to the data rpovider
* the data provider is part of the protocol spec
* merge in the beacon and block hash processing into the same class
* MainnetProtocolSpecs is where it all gets loaded in.